exotica_core/src/visualization_meshcat.cpp

```diff
@@ -0,0 +1,342 @@
+    ret[0] = frame.M.data[0] * scale_x;
+    ret[1] = frame.M.data[3];
+    ret[2] = frame.M.data[6];
+    ret[3] = 0.0;
+    ret[4] = frame.M.data[1];
+    ret[5] = frame.M.data[4] * scale_y;
+    ret[6] = frame.M.data[7];
+    ret[7] = 0.0;
+    ret[8] = frame.M.data[2];
+    ret[9] = frame.M.data[5];
+    ret[10] = frame.M.data[8] * scale_z;
+    ret[11] = 0.0;
+    ret[12] = frame.p.data[0];
+    ret[13] = frame.p.data[1];
+    ret[14] = frame.p.data[2];
+    ret[15] = 1.0;
+    return ret;
+};
+
+std::vector<double> PositionToVector(const KDL::Frame& frame)
+{
+    std::vector<double> ret(3);
+    ret[0] = frame.p.data[0];
+    ret[1] = frame.p.data[1];
+    ret[2] = frame.p.data[2];
+    return ret;
+};
+
+std::vector<double> QuaternionToVector(const KDL::Frame& frame)
+{
+    std::vector<double> ret(4);
+    frame.M.GetQuaternion(ret[0], ret[1], ret[2], ret[3]);
+    return ret;
+};
+
+VisualizationMeshcat::VisualizationMeshcat(ScenePtr scene, const std::string& url, bool use_mesh_materials) : scene_(scene), context_(1), zmq_url_(url)
+{
+    HIGHLIGHT_NAMED("VisualizationMeshcat", "Initialising visualizer");
+    Initialize(use_mesh_materials);
+}
+VisualizationMeshcat::~VisualizationMeshcat() = default;
+
+void VisualizationMeshcat::Initialize(bool use_mesh_materials)
+{
+    ConnectZMQ();
+    web_url_ = RequestWebURL();
+    if (web_url_.size() > 7) file_url_ = web_url_.substr(0, web_url_.size() - 7) + "files/";
+    ConnectZMQ();
+    path_prefix_ = "/exotica/" + scene_->GetName() + "/";
+}
+
+void VisualizationMeshcat::ConnectZMQ()
+{
+    socket_ = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(context_, ZMQ_REQ));
+    socket_->setsockopt(ZMQ_RCVTIMEO, TIMEOUT);
+    socket_->connect(zmq_url_);
+}
+
+void VisualizationMeshcat::SendZMQ(const std::string& data)
+{
+    zmq::message_t request(data.size());
+    std::memcpy(request.data(), data.c_str(), data.size());
+    socket_->send(request);
+}
+
+std::string VisualizationMeshcat::ReceiveZMQ()
+{
+    char buffer[2048];
+    std::memset(buffer, 0, 2048);
+    socket_->recv(&buffer, 2048);
+    return std::string(buffer);
+}
+
+std::string VisualizationMeshcat::RequestWebURL()
+{
+    SendZMQ("url");
+    return ReceiveZMQ();
+}
+
+std::string VisualizationMeshcat::GetWebURL()
+{
+    return web_url_;
+}
+
+std::string VisualizationMeshcat::GetFileURL()
+{
+    return file_url_;
+}
+
+template <typename T>
+void VisualizationMeshcat::SendMsg(T msg)
+{
+    msgpack::sbuffer sbuf;
+    msgpack::pack(sbuf, msg);
+    socket_->send(msg.type.data(), msg.type.size(), ZMQ_SNDMORE);
+    socket_->send(msg.path.data(), msg.path.size(), ZMQ_SNDMORE);
+    socket_->send(sbuf.data(), sbuf.size());
+    ReceiveZMQ();
+}
+
+void VisualizationMeshcat::DisplayScene(bool use_mesh_materials)
+{
+    const std::vector<std::weak_ptr<KinematicElement>>& elements = scene_->GetKinematicTree().GetTree();
+
+    for (std::weak_ptr<KinematicElement> weak_element : elements)
```

Comment:
You can do the link in the for-standard if you want

Reply:
Do you mean the `lock()`?
I don't mind the extra line.